### ISSUE TITLE
Building identical images results in different digests, even with `--reproducible`

### ISSUE THREAD

**Issue by GijsWeterings** [GoogleContainerTools/kaniko#710]
I'm building images using kaniko on my kubernetes cluster, but am seeing some weird results with respect to reproducibility.

**Actual behavior**
The digest seems to be one of two values

**Expected behavior**
Every time the same image is built, it should have the exact same digest

**To Reproduce**

1. run kaniko with `--reproducible --no-cache ` 30x
2. observe different digests, in my case `sha256:a252f9d8e2f5078e2010cdcc6ce377201e5918c6932e5edb049939b12f8a0e38` and `sha256:7ee8806db7d788db7e23462c1f8df30f8ad814c2c4b486f75420c08fbb8c3bb1`
3. use `dive` to compare images, see the layer of the `RUN` statement produced a different layer sha [screenshots](https://imgur.com/a/Emc7gEz)
4. Boot both images, see `sha256sum test.log` gives the same hash
5. `stat` both files, see [screenshots](https://imgur.com/a/XruOfwi) for difference. I don't know exactly what constitutes the sha of a layer, but notices some of the timestamps and the inode count are different here (despite `--reproducible`)

**Additional Information**
 - Dockerfile 
```
FROM busybox@sha256:bf510723d2cd2d4e3f5ce7e93bf1e52c8fd76831995ac3bd3f90ecc866643aff
RUN echo "Hi" > test.log
```
 - Kaniko Image (fully qualified with digest)
gcr.io/kaniko-project/executor@sha256:78d44ec4e9cb5545d7f85c1924695c89503ded86a59f92c7ae658afa3cff5400

Do you have any clue on why this could be different? I noticed no matter how often I run this, the two hashes both keep occuring and both stay the same. Below I also attached build logs of a run of each build. the count in the logrus INFO[XXXX] seems to differ between builds of the same sha as well

build log of run producing the sha256:7ee8806db7d788db7e23462c1f8df30f8ad814c2c4b486f75420c08fbb8c3bb1 image:
```                                  
[36mINFO[0m[0001] Resolved base name registry2.kubernetes.gijsweterings.nl/library/busybox@sha256:bf510723d2cd2d4e3f5ce7e93bf1e52c8fd76831995ac3bd3f90ecc866643aff to registry2.kubernetes.gijsweterings.nl/library/busybox@sha256:bf510723d2cd2d4e3f5ce7e93bf1e52c8fd76831995ac3bd3f90ecc866643aff 
[36mINFO[0m[0001] Resolved base name registry2.kubernetes.gijsweterings.nl/library/busybox@sha256:bf510723d2cd2d4e3f5ce7e93bf1e52c8fd76831995ac3bd3f90ecc866643aff to registry2.kubernetes.gijsweterings.nl/library/busybox@sha256:bf510723d2cd2d4e3f5ce7e93bf1e52c8fd76831995ac3bd3f90ecc866643aff 
[36mINFO[0m[0001] Error while retrieving image from cache: getting file info: stat /cache/sha256:bf510723d2cd2d4e3f5ce7e93bf1e52c8fd76831995ac3bd3f90ecc866643aff: no such file or directory 
[36mINFO[0m[0001] Downloading base image registry2.kubernetes.gijsweterings.nl/library/busybox@sha256:bf510723d2cd2d4e3f5ce7e93bf1e52c8fd76831995ac3bd3f90ecc866643aff 
[36mINFO[0m[0002] Built cross stage deps: map[]                
[36mINFO[0m[0002] Error while retrieving image from cache: getting file info: stat /cache/sha256:bf510723d2cd2d4e3f5ce7e93bf1e52c8fd76831995ac3bd3f90ecc866643aff: no such file or directory 
[36mINFO[0m[0002] Downloading base image registry2.kubernetes.gijsweterings.nl/library/busybox@sha256:bf510723d2cd2d4e3f5ce7e93bf1e52c8fd76831995ac3bd3f90ecc866643aff 
[36mINFO[0m[0002] Unpacking rootfs as cmd RUN echo "Hi" > test.log requires it. 
[36mINFO[0m[0003] Taking snapshot of full filesystem...        
[36mINFO[0m[0006] RUN echo "Hi" > test.log                     
[36mINFO[0m[0006] cmd: /bin/sh                                 
[36mINFO[0m[0006] args: [-c echo "Hi" > test.log]              
[36mINFO[0m[0006] Taking snapshot of full filesystem...        
2019/07/03 10:16:04 existing blob: sha256:b0164d3ecf44fce4d745b75b4e2a1cfd763260aefbece8c7c7f88619bfd4a583
2019/07/03 10:16:04 existing blob: sha256:80e505540c72927030882ec589b874017bc5fa2184ac448539ab4f22c0778d32
2019/07/03 10:16:04 existing blob: sha256:f22a7de439f0a30c07b6a8b7a9af4faca21d4fcf33f7b6b3f97c846fe4aa2873
2019/07/03 10:16:09 registry2.kubernetes.gijsweterings.nl/library/hellodocker:4fdb38180f08dcb16832412bbf6fb52cca4f31c4: digest: sha256:7ee8806db7d788db7e23462c1f8df30f8ad814c2c4b486f75420c08fbb8c3bb1 size: 587

```

sha256:a252f9d8e2f5078e2010cdcc6ce377201e5918c6932e5edb049939b12f8a0e38:
```                               
[36mINFO[0m[0000] Resolved base name registry2.kubernetes.gijsweterings.nl/library/busybox@sha256:bf510723d2cd2d4e3f5ce7e93bf1e52c8fd76831995ac3bd3f90ecc866643aff to registry2.kubernetes.gijsweterings.nl/library/busybox@sha256:bf510723d2cd2d4e3f5ce7e93bf1e52c8fd76831995ac3bd3f90ecc866643aff 
[36mINFO[0m[0000] Resolved base name registry2.kubernetes.gijsweterings.nl/library/busybox@sha256:bf510723d2cd2d4e3f5ce7e93bf1e52c8fd76831995ac3bd3f90ecc866643aff to registry2.kubernetes.gijsweterings.nl/library/busybox@sha256:bf510723d2cd2d4e3f5ce7e93bf1e52c8fd76831995ac3bd3f90ecc866643aff 
[36mINFO[0m[0000] Error while retrieving image from cache: getting file info: stat /cache/sha256:bf510723d2cd2d4e3f5ce7e93bf1e52c8fd76831995ac3bd3f90ecc866643aff: no such file or directory 
[36mINFO[0m[0000] Downloading base image registry2.kubernetes.gijsweterings.nl/library/busybox@sha256:bf510723d2cd2d4e3f5ce7e93bf1e52c8fd76831995ac3bd3f90ecc866643aff 
[36mINFO[0m[0001] Built cross stage deps: map[]                
[36mINFO[0m[0001] Error while retrieving image from cache: getting file info: stat /cache/sha256:bf510723d2cd2d4e3f5ce7e93bf1e52c8fd76831995ac3bd3f90ecc866643aff: no such file or directory 
[36mINFO[0m[0001] Downloading base image registry2.kubernetes.gijsweterings.nl/library/busybox@sha256:bf510723d2cd2d4e3f5ce7e93bf1e52c8fd76831995ac3bd3f90ecc866643aff 
[36mINFO[0m[0003] Unpacking rootfs as cmd RUN echo "Hi" > test.log requires it. 
[36mINFO[0m[0003] Taking snapshot of full filesystem...        
[36mINFO[0m[0006] RUN echo "Hi" > test.log                     
[36mINFO[0m[0006] cmd: /bin/sh                                 
[36mINFO[0m[0006] args: [-c echo "Hi" > test.log]              
[36mINFO[0m[0006] Taking snapshot of full filesystem...        
2019/07/03 10:15:56 existing blob: sha256:841059053790a500197d8ff598f8e0fafe51211160627765bfece7e432a6861a
2019/07/03 10:15:56 existing blob: sha256:f22a7de439f0a30c07b6a8b7a9af4faca21d4fcf33f7b6b3f97c846fe4aa2873
2019/07/03 10:15:56 existing blob: sha256:78e9d4f42ddfe7875b3033ce1fee3467a192680c33abe82af9a07852130cd31d
2019/07/03 10:16:01 registry2.kubernetes.gijsweterings.nl/library/hellodocker:4fdb38180f08dcb16832412bbf6fb52cca4f31c4: digest: sha256:a252f9d8e2f5078e2010cdcc6ce377201e5918c6932e5edb049939b12f8a0e38 size: 587

```

**Comment by priyawadhwa** [comment]
Fixed by #714, closing this issue. Please reopen/comment if you're still seeing this problem!

**Comment by mjgallag** [comment]
@priyawadhwa I'm seeing this issue using `--cache=true --reproducible=true` with `gcr.io/kaniko-project/executor:d4badf36d67a8fc4c3e38162bf1b697e20d55c69` on GCB. It also occurs when I don't include `--reproducible=true`, which I thought might fix it. Note this is also an issue with `gcr.io/kaniko-project/executor:v0.13.0` but I figured I'd try it with master branch executor too. My cache layers all produce hits and are reused with each build but the final image always has a different digest. I've tried a few different images and get the same result. Here is a very simple one that reproduces the issue:
```
FROM node:10.17.0

RUN npm install npm@6.12.0 --global && npm cache clean --force
```

**Comment by priyawadhwa** [comment]
@mjgallag thanks for reporting this! If you, or anyone else, would be interested in exploring this bug, please let me know and I can assign you to this issue.

**Comment by mjgallag** [comment]
@priyawadhwa thanks for re-opening! I just wanted to note that I discovered this project from https://cloud.google.com/cloud-build/docs/kaniko-cache which suggests it as a solution to the lack of a distributed docker cache in Google Clould Build. I think it would still probably be worth it to use even with this issue if you have no dependencies between your Docker images. Unfortunately, we have Docker images that are built FROM other Docker images of ours and this causes no cache hits to occur for our dependent images.

**Comment by ingokofler** [comment]
I also experience an issue with having non-reproducible layers even when using `--reproducible` flag set. Note that I do not use caching ATM.

My setup is as follows - I build a Docker image with a Spring Boot application (app.jar). Before starting kaniko I unzip the jar as follows

```
unzip app.jar -d exploded
```
The Dockerfile looks as follows
```
ARG DEBUG_SUFFIX
FROM gcr.io/distroless/java:8$DEBUG_SUFFIX

# explicit switch to nonroot user
USER nonroot

# see SpringBoot documentation for details
# https://spring.io/guides/topicals/spring-boot-docker/

COPY exploded/BOOT-INF/lib /app/lib
COPY exploded/META-INF /app/META-INF
COPY exploded/BOOT-INF/classes /app
```
The intention was to have the dependencies in a separate layer that remains stable so should end up with the same digest. Nevertheless, when running this multiple times in the CI one get different layer digests. When having a closer look I noticed that the tar-split.json.gz files are different in a sense that they have the same files in it but in a different order.

BTW I am using kaniko v0.13.0. Also worth noting that I first had a multi-stage build to extract the content of the jar but this lead to the same result.

**Comment by cvgw** [comment]
closing this in favor of continued tracking in #900
